dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala

```diff
@@ -29,83 +29,172 @@ class Recurrent[T : ClassTag] (
   bpttTruncate: Int = 2)
   (implicit ev: TensorNumeric[T]) extends Container[Tensor[T], Tensor[T], T] {
 
-  val hidden = Tensor[T]()
-  var module: Module[T] = _
-  var transform: Module[T] = _
+  var hidden: Activity = null
+  val inputs = T()
+  val batchDim = 1
+  val timeDim = 2
+  val hidDim = 1
+  val inputDim = 2
   var (batchSize, times) = (0, 0)
 
+  override def reset(): Unit = {
+    modules(0).reset()
+    modules(0).isInstanceOf[RnnCell[T]] match {
+      case true => hidden = Tensor[T]()
+      case _ => throw new IllegalArgumentException("Cell not implemented")
+    }
+  }
+
+  /**
+   * Clone N models; N depends on the time dimension of the input
+   * @param size
+   */
+  private def extend(size: Int): Unit = {
+    require(modules != null && modules.length >= 1,
+      "Recurrent extend: should contain at least one module")
+    var t = modules.length
+    var flag = false
+    if (t < size) {
+      flag = true
+    }
+    while (t < size) {
+      super.add(modules(0).cloneModule())
+      t += 1
+    }
+    if (flag) {
+       share()
+    }
+  }
+
+  /**
+   * Sharing weights, gradWeights across all the modules in time dim
+   */
+  def share(): Unit = {
+    val params = modules(0).parameters()
+    var i = 1
+    while (i < modules.length) {
+      val curParams = modules(i).parameters()
+      var j = 0
+      while (j < curParams._1.length) {
+        curParams._1(j).set(params._1(j))
+        j += 1
+      }
+      j = 0
+      while (j < curParams._2.length) {
+        curParams._2(j).set(params._2(j))
+        j += 1
+      }
+      i += 1
+    }
```

Comment:
Can we do something similar to https://github.com/intel-analytics/BigDL/blob/master/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala#L386